src/redis_memory.rs

```diff
@@ -0,0 +1,242 @@
+    pub fn flush_db(&self) -> Result<(), redis::RedisError> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        redis::cmd("FLUSHDB").exec(safe_connection)?;
+        Ok(())
+    }
+}
+
+/**
+ * The RedisMemory implementation of the MemoryADT trait.
+ * All operations are - and should be - atomic.
+ */
+impl<Address: Send + Sync + Hash + Eq + Debug + Clone + ToRedisArgs, const WORD_LENGTH: usize>
+    MemoryADT for RedisMemory<Address, WORD_LENGTH>
+{
+    type Address = Address;
+    type Error = RedisMemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    /**
+     * Atomically reads the values at the given addresses.
+     */
+    async fn batch_read(
+        &self,
+        addresses: Vec<Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        let refs: Vec<&Address> = addresses.iter().collect(); // Redis MGET requires references to the values
+        let res: Vec<Option<Self::Word>> =
+            safe_connection.mget(&refs).map_err(|e| RedisMemoryError {
+                details: e.to_string(),
+            })?;
+        Ok(res)
+    }
+
+    /**
+     * Atomically writes the bindings if the guard is still valid.
+     * Returns the current value on the guard's address.
+     * If the result is equal to the guard's old value, the bindings get
+     * written.
+     *
+     * Args that are passed to the LUA script are, in order :
+     * 1. Guard address.
+     * 2. Guard value.
+     * 3. Vector length.
+     * 4+. Vector elements (address, word).
+     */
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        bindings: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<Self::Word>, Self::Error> {
+        let safe_connection = &mut *self.connection.lock().expect("Poisoned lock.");
+        let (guard_address, guard_value) = guard;
+
+        const GUARDED_WRITE_LUA_SCRIPT: &str = r#"
+        local guard_address = ARGV[1]
+        local guard_value = ARGV[2]
+        local length = ARGV[3]
+        
+        local value = redis.call('GET',ARGV[1])
+
+        -- compare the value of the guard to the currently stored value
+        if((value==false) or (not(value == false) and (guard_value == value))) then
+            -- guard passed, loop over bindings and insert them
+            for i = 4,(length*2)+3,2
+            do
+                redis.call('SET', ARGV[i], ARGV[i+1])
+            end
+        end
+        return value
+        "#;
+
+        let script = redis::Script::new(GUARDED_WRITE_LUA_SCRIPT);
+        let mut script_invocation = script.prepare_invoke();
+
+        script_invocation.arg(guard_address);
+        if let Some(byte_array) = guard_value {
+            script_invocation.arg(&byte_array);
+        } else {
+            script_invocation.arg("false".to_string());
+        }
```

Comment:
Not idiomatic. Use `unwrap_or` or something similar.

Reply:
```rust
        script_invocation.arg(&guard_value.unwrap_or("false".to_string()));
```

I beleive this is the idiomatic way of doing it, however it fails compilation because rust expects the same type for the "unwrap_or". I didn't find a shorter way

Reply:
You may be right, I didn't notice the types weren't the same. There may be a solution using dynamic dispatch but it is ugly. Keep it as is.